### ISSUE TITLE
Display album art instead of fb2k logo

### ISSUE THREAD

**Issue by popeyaho001** [TheQwertiest/foo_discord_rich#6]
As the title, can we show the image album instead foobar icon?

**Comment by TheQwertiest** [comment]
Unfortunately, I can do nothing about - it's a limitation of Discord API.
I wish I could make this plugin look the same as Spotify integration, but, alas, Discord developers are not yet ready to give such tools to 3rd parties =(

**Comment by TheQwertiest** [comment]
Relevant links on Discord boards:
https://support-dev.discordapp.com/hc/en-us/community/posts/360043641772-Allow-use-of-progress-bar-links-in-activity
https://support.discordapp.com/hc/en-us/community/posts/360053661011-Allow-different-Rich-Presence-configurations-over-IPC-API

**Comment by TheLastZombie** [comment]
For anyone who is also trying to hack their way around this, rich presence assets can be uploaded via a `POST` request:

```
curl -X POST https://discordapp.com/api/v6/oauth2/applications/<CLIENT ID>/assets \
     -H "Authorization: <TOKEN>" \
     -H "Content-Type: application/json" \
     -d "{'image':' data:image/png;base64,<BASE64 IMAGE>','name':'<ASSET NAME>','type':1}"
```

When sending the request, Discord returns a JSON object which includes an ID, which in turn can be used to delete the asset again, though I haven't tried that yet. Apparently, you need to send a DELETE request to `https://discordapp.com/api/v6/oauth2/applications/<CLIENT ID>/assets/<ASSET ID>`.

Some things to note:
- The authorization token seems to be different from the one you can get via the Discord application. I got it by uploading an asset via Firefox and inspecting the request afterwards.
- Obviously, the MIME type would need to be adjusted for non-PNGs.
- All RPC assets (so, in this case, all album covers) need to have a resolution of at least 512x512 px.
- When viewing the art assets in a browser, Discord may show assets already deleted or not show recently uploaded ones. The API they use to list those assets has some serious caching issues...

**Comment by Galixte** [comment]
Hi @TheQwertiest,

is it possible to add an option to set a customize image logo of Foobar2k instead of the original?

Regards.

**Comment by TheQwertiest** [comment]
No, see https://github.com/TheQwertiest/foo_discord_rich/issues/6#issuecomment-467542756

**Comment by Mimi1001** [comment]
The Discord Rich Presence Plug-In for the music player MusicBee, [DiscordBee](https://github.com/sll552/DiscordBee), recently introduced support for custom album art by registering a custom DiscordApp and automatically uploading covers as assets. Would it be possible to implement the feature similarly for this foobar2000 Plug-In?

**Comment by TwoZeroes** [comment]
Any hopes of this Issue being opened up again? I hope to see this feature added.

**Comment by Otuxam3** [comment]
I'm just sending this as a good reason to reopen the Issue.

As I have seen on Reddit, someone managed to use an image link as large asset image, so I wanted to do it myself with a custom RPC client to be sure this wasn't another fake, and I can tell that it indeed works. It can be easily reproduced on your side.
(some infos hidden for privacy)

![console](https://user-images.githubusercontent.com/58670174/147368106-b344a11e-a6ea-424d-bedc-cc71bb70261b.png)
![result](https://cdn.discordapp.com/attachments/734115770488848414/923916542255693855/Screenshot_20211224-133412_Discord-Beta.jpg)

For now there's no official documentation, but it's something to be seen soon enough in the future.
It would also be a good reason to allow foobar title formatting in the assets tab.

Our prayers have been heard guys, API isn't anymore limiting us 😄

**Comment by TheQwertiest** [comment]
@Otuxam3 that's good news :)
But I can't use it, since it's not a part of official API (yet?)

**Comment by Otuxam3** [comment]
I initially wanted to edit my message at the moment the PR appeared but preferred to send a followup here.
So by watching API docs repo every single day this is about to be merged, just a matter of days now 👌🏻.
https://github.com/discord/discord-api-docs/pull/4317

So, wait and see I'd say ^^

**Comment by LLinoor** [comment]
is there any news?

**Comment by Otuxam3** [comment]
Everything is in place Discord-side, PR merge in the repo has been reflected on the [website](https://discord.com/developers/docs/topics/gateway#activity-object-activity-assets).
I just don't know if this is enough for @TheQwertiest to start implementing it since this part of the docs only informs about an activity object as sent by the gateway.
RPC docs do redirect to that page ([follow the links](https://discord.com/developers/docs/topics/rpc#setactivity)).
As for SDK, I don't use it when doing my stuff since I have no Idea what to change to use it via node-gyp so I stick with RPC only 😅 

But my job here is done, it's now your turn to decide what to do, dear maintainer! :)

**Comment by thinksystem** [comment]
I was going to start a issue about this but looks like people already thought of wanting to set the album art as the rich presence image instead of the foobar2000 logo.

Can't wait for when the maintainer implements this feature 😄

**Comment by tort-oise** [comment]
(Don't mind the "playing image (dark)" section, I was messing around and doing that actually makes it look like this:
![image](https://user-images.githubusercontent.com/81963124/162969742-8333ef63-76c4-4bda-8cdf-b84071e6b1fd.png)
Doubles the image. Not recommended, unless you're a huge fan.

**Comment by Andreas711** [comment]
I'm sorry if I'm waking a necro thread but I'm sure somebody could easy whip up a script that pulls album covers from fb2k and then set the playing icon to said album cover.

**Comment by BobbyWibowo** [comment]
The PoC in @tort-oise's comment appears to be the solution that is used in [phineas05](https://github.com/phineas05) /
[discord-rich-presence-plex](https://github.com/phineas05/discord-rich-presence-plex), **but automated**

It automatically uploads the currently playing movie's poster into imgur (if it hasn't yet been uploaded, otherwise use cache), then use the direct link as-is for `large_image` asset

References:
https://github.com/phineas05/discord-rich-presence-plex/blob/da7d53d1d698642095559bd4be00930267e869cc/services/PlexAlertListener.py#L203-L219
https://github.com/phineas05/discord-rich-presence-plex/blob/da7d53d1d698642095559bd4be00930267e869cc/services/imgur.py

Discord will automatically proxy the image afterwards without the app having to explicitly do so first:
![image](https://i.fiery.me/36fC.png)

> There is a reference to `images.weserv.nl` in the screenshot above because I added [that proxy](https://images.weserv.nl/docs/) in my own [personal-use fork](https://github.com/BobbyWibowo/discord-rich-presence-plex/commit/c746fb153f75f68362959c75fdf095063d77909b)
> Using the full image as-is did not feel too ideal when they're only being rendered as 90x90 square at most, so I'd prefer it to go through the proxy first for some automated resizing and cropping
> But yeah, it appears Discord does not particularly care from which the remote image comes from, and will just automatically proxy them either way

**Comment by degu-Tanya** [comment]
bump, as the only still worked on foobar2000 discord presence I could find on google, it would be a nice addition.

**Comment by DJTOMATO** [comment]
Bump

**Comment by thatonefaker** [comment]
why is this closed now?

**Comment by LeviOP** [comment]
it was closed a long time ago, hasn't been reopened because the owner presumably doesn't feel like programming this right now

**Comment by s0hv** [comment]
I've been working on a prototype for this and I've created something that works in the simplest cases. It can be found on my fork [here](https://github.com/s0hv/foo_discord_rich/tree/cover-art-images). You need to manually build it to try it out if you want. I would recommend using a backup instance of foobar instead of your main one if you want to try it out. 

Here's an image of the change in action
![image](https://user-images.githubusercontent.com/17063236/184863614-8cec9083-f83f-4ec6-932d-0936e4425787.png)

This implementation leaves the image uploading to the user, as I didn't want to lock people to one vendor. Also because I use my own image host. The system I wrote just calls a user specified command (e.g. an exe or a bat file) and writes the image filepath to the processes stdin. It then tries to read the output of the process and use that as the url. Here's the shell script I used during testing, which uses ShareX for uploading the image.

```sh
myvar=`cat`

"C:\Program Files\ShareX\ShareX.exe" "$myvar" -s -autoclose -m -nohotkeys
cat /dev/clipboard
```
I run it with git bash but any shell with `cat` should suffice. The command looks something like this 
`"C:\Program Files\Git\bin\bash.exe" "C:\path\to\upload.sh"`
This command should be used in the plugin configuration on the advanced tab, which now has a new section for this feature.
![image](https://user-images.githubusercontent.com/17063236/184865389-a08382a8-d64d-4928-aea1-61d1f766a470.png)

I can finish this feature using this approach but would be nice to get some feedback from @TheQwertiest as I don't know
if this is how they'd want it to work. Also because the code quality might suck since I haven't used C++ in a long time and never written a foobar plugin before.

Maybe this issue should be reopened now since it is possible to make it work.

**Comment by BDtetra** [comment]
maybe an md5 hash check as well to see if the album art needs an update in the first place before actually uploading the new image (since often times you would listen to an album, where the art doesnt change).  Maybe this is taking a little too far but it would help to also open a port to host the image on your ip(through a desired port),

**Comment by s0hv** [comment]
I've thought about the md5 hashing thing but I am unsure as to how I'd implement the storage of it. I created a new PR #37 for this so it might be best to move the discussion there relating to my implementation.

**Comment by prism2001** [comment]
@s0hv how do you compile this? i have visual studio but i've never built a C++ project (or anything off github for that matter) so i'm completely lost. i tried cloning it and building the solution but it threw a bunch of errors, so build instructions would be appreciated for those of us who want to try it out but aren't already developers

**Comment by s0hv** [comment]
@prism2001 There are automatic AppVeyor builds done every time my PR #37 gets new commits pushed so you snatch one from there. Just use it in a copy of foobar as the plugin might break stuff between updates.

**Comment by topecnz** [comment]
![Discord_hpFg9OkkNf](https://user-images.githubusercontent.com/14878600/186705665-0867d1bd-f6bf-4051-8812-37cd45dc3c9c.gif)

It took 2-3 days to make but it seems the codes are still messy or lacking some stuffs to add. The problem is, I don't know on how the GUID works for C++.

By the way, I used lastfm api to get the album cover and externally fetched from python. I tried curl and cpr with jsoncpp but something went wrong when compiling. For now, I am using popen to execute (I'm dealing some opening windows when changing the song.)

**Comment by prism2001** [comment]
@s0hv

got it working and so far it's been really nice. a few questions though:

1. where is the JSON file stored? the one that's supposed to keep all the hashes and URLs for cover art. i'm using your provided imgur sample python script
2. when i was using your sharex shell script, i had briefly disabled the option in sharex to automatically copy the URL of the artwork, which for some reason prevented the artwork from appearing in discord until i turned it back on, but for the albums that had their artwork uploaded when i had the URL copying disabled, they now only show up as a generic icon, and i'm not sure how to force the plugin to forget these albums or force a proper re-upload. any advice on this?

**Comment by s0hv** [comment]
@prism2001 The json file is stored in your foobar2000 folder in a folder called foo_discord_rich. It might be a few folders deep depending on your version and install type. To clear the urls just select all the affected tracks, right click and select Discord rich presence -> Clear artwork. You can open the console to see logs related to the plugins behaviour.

**Comment by prism2001** [comment]
@s0hv 
I've checked the folder and there doesn't appear to be any JSON file. There are two of these folders, actually. One in _AppData\Roaming\foobar2000\foo_discord_rich_ and another one in _AppData\Roaming\foobar2000\user-components\foo_discord_rich_ - the first one is completely empty, and the other one is just where the plugin DLL is located, along with the licenses

**Comment by s0hv** [comment]
@prism2001 The file is created only after an image is actually uploaded with the latest version and the file will probably be created in the first folder.

**Comment by prism2001** [comment]
@s0hv 

i'm using the latest version that was built by appveyor (5 days ago) and it has uploaded numerous images so far, but nothing is being saved to a JSON on my end

**Comment by s0hv** [comment]
@prism2001 Try checking the console logs. When it's trying to upload an image again it should say that a hash was found or if it was not found it should say what the upload script output. Otherwise I'm not sure where the file is saved on normal installations. I use a portable installation myself.

**Comment by prism2001** [comment]
@s0hv

upon booting foobar and loading any track, the only thing that appears in the console regarding the plugin is that it connected to my discord successfully. when I play tracks from albums that have already had their artworks uploaded nothing pops up at all in the console. the only time i get any other output from the plugin in the console is whenever I load an album that hasn't already had its cover art uploaded - the output for those being what I posted in the screenshot above

**Comment by s0hv** [comment]
@prism2001 it's working as intended then. Try clearing the artwork and reuploading it (might have to do it twice). If the hashing is working correctly it should find the old url when you try to re-upload the image.

**Comment by s0hv** [comment]
@prism2001 Finally was able to look at the code and it seems that I had forgotten to test the code when the json file has not been created. I did a fix and now the json file should be created correctly in that situation. The hashes won't be created for already uploaded artwork, but that shouldn't really be an issue in most cases.

**Comment by LastReaction** [comment]
@s0hv I have to say, thank you very much for adding this must-have feature to this repo! Everything for the most part seems to be working great on my end using the Imgur API Python script. I am running into one small issue where certain album covers won't upload to Imgur, however. I'm getting a generic: "File is already in use" error when trying to generate the Imgur URL manually in foobar2000. It only seems to happen on specific album covers, but works perfectly fine with majority of my music library.

**Comment by s0hv** [comment]
@GH0STDATA The script handles album art differently depending on if it is embedded in the music file or if it is a separate file on the disk. For embedded files the image is copied to disk and that temporary file is used. If the file is already on disk that path is used. It's likely that this difference is what causes the "File is already in use" error.

**Comment by llama8** [comment]
How can I tell your addon to pick a front.jpg album art located in current playing song directory?

**Comment by SimonFoxcoon** [comment]
s0hv's fork, is there a way to make the component check for per-album or per-track album covers? they will default to some album cover, usually the first one i play, and never changes if i manually clear it.

**Comment by s0hv** [comment]
@SimonFoxcoon 
> s0hv's fork, is there a way to make the component check for per-album or per-track album covers? they will default to some album cover, usually the first one i play, and never changes if i manually clear it.

"Artwork identifier key" might be what you're looking for
https://s0hv.github.io/foo_discord_rich/docs/album-artwork/#descriptions-of-options

**Comment by TobyTobyReal** [comment]
I've been messing with the artwork identifier key for hours now trying to make it show per-track album covers, can someone help please?

**Comment by s0hv** [comment]
> I've been messing with the artwork identifier key for hours now trying to make it show per-track album covers, can someone help please?
@TobyTobyReal Would this work `%album artist% [- %album%] - %title% [- %discnumber%]`

**Comment by TobyTobyReal** [comment]
> > I've been messing with the artwork identifier key for hours now trying to make it show per-track album covers, can someone help please?
> > @TobyTobyReal Would this work `%album artist% [- %album%] - %title% [- %discnumber%]`

Unfortunately that doesnt seem to work either, thanks for trying to help though!

**Comment by shirafukayayoi** [comment]
[s0hv](https://github.com/s0hv)
I am Japanese and am not good at English, so I use translation software.
 My question is, what is the name of the metadata that contains the album art of imgur used in this s0hv plugin?
 %title% %artist%
 It's a part like that.

**Comment by shirafukayayoi** [comment]
@s0hv 
I am Japanese and am not good at English, so I use translation software.
 My question is, what is the name of the metadata that contains the album art of imgur used in this s0hv plugin?
 %title% %artist%
 It's a part like that.

Addendum.

What I'm thinking of doing now is putting the playback information into a txt file with the nowmusic plugin, reading it and displaying it as DIscordRIchPresence.
https://github.com/shirafukayayoi/Node_fb2kRichPresence
I want the nowmusic plugin to output the name of the imgur file it creates, but I can't do that because it's not metadata.

Do you have any ideas?

**Comment by s0hv** [comment]
@shirafukayayoi Unfortunately I don't think what you're trying to do is possible. The plugin does not expose any metadata fields that can be read with the %xxxxxxx% syntax.

**Comment by TheQwertiest** [comment]
@s0hv thanks for your research and work on image upload!
If you don't mind, I might use your code as inspiration (with credits and stuff) once I'm ready to add it.

For now though, I'm working on fetching image urls from MusicBrainz, so that users might have some art without having to upload it.

**Comment by s0hv** [comment]
@TheQwertiest Yeah go ahead. Glad to see this finally being implemented in some way 😁

**Comment by kam106** [comment]
I don't mean to necro this issue thread, but I would personally enjoy if this kind of rich presence also had text that would appear when you hover over the Large Image Key. One of Musicbee's plugins DiscordBee supports this, and I think it would be a good addition to this kind of feature.

Here's an image reference of this in action:
![Discord_PGKBDsTu7Y](https://github.com/user-attachments/assets/a345a1b5-cc41-4edb-92b7-7836a1833104)
